File: leetcode/subarray-sum-equals-k.py
# https://leetcode.com/problems/subarray-sum-equals-k/

# Tags: Array - Hash Table - Prefix Sum

import timeit
from collections import defaultdict
from typing import List


# Create a dictionary to store the prefix sums we have already seen and how many times we have seen each.
# When we calculate each prefix sum we check if the current prefix sum - our target is in the prefix sums
# we have seen. If found, we add its value, the number of times we have seen that prefix sum, to the result
# count. Then we add the prefix sum entry to the dictionary, or add to the previous count if previously seen.
#
# Sum(i,j) = Sum(0,j) - Sum(0,i) => Sum(0,i) = Sum(0,j) - Sum(i,j)
#
# Time complexity: O(n) - we iterate over the elements of the array once and the dictionary once 2*n = O(n)
# checking if the difference is in the dictionary is O(1).
# Space complexity: O(n) - the dictionary will grow to size n.
#
# Runtime: 324 ms, faster than 85.43% of Python3 online submissions for Subarray Sum Equals K.
# Memory Usage: 16.6 MB, less than 32.25% of Python3 online submissions for Subarray Sum Equals K.
class PrefixSum:
    def subarraySum(self, nums: List[int], k: int) -> int:
        sums = defaultdict(int)
        sums[0] = 1

        # Calculate all prefix sums
        res, sum = 0, 0
        for num in nums:
            # Calculate the prefix sum at this element
            sum += num

            # If we can get k with this element and one, or multiple, previous ones, add that to the result set
            if sum - k in sums:
                res += sums[sum - k]

            # Add this prefix sum to the sums dictionary
            sums[sum] += 1

        return res


def test():
    executors = [PrefixSum]
    tests = [
        [[1, 1, 1], 2, 2],
        [[1, 2, 3], 3, 2],
        [[1], 0, 0],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(int(float("1"))):
            for i, t in enumerate(tests):
                sol = executor()
                result = sol.subarraySum(t[0], t[1])
                exp = t[2]
                assert (
                    result == exp
                ), f"\033[93m» {result} <> {exp}\033[91m for test {i} using \033[1m{executor.__name__}"
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        res = "{0:20}{1:10}{2:10}".format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/subarray-sum-equals-k.rs
// 560. Subarray Sum Equals K
// 🟠 Medium
//
// https://leetcode.com/problems/subarray-sum-equals-k/
//
// Tags: Array - Hash Table - Prefix Sum

use std::collections::HashMap;

struct Solution;
impl Solution {
    /// Iterate over the array keeping the sum of values up to the current index,
    /// we store all sums that we have seen in a hashmap of sum => number of
    /// times that we have seen that sum. At each index, we check if we can find
    /// the value of the current sum - k in the hashmap, if we can, we add the
    /// number of times that we have seen it to the result, because the subarray
    /// between the current index and the indexes with s - k have a sum that
    /// equals k.
    ///
    /// Time complexity: O(n) - We iterate over the input array once and do
    /// constant time work for each element.
    /// Space complexity: O(n) - The hashmap will have an element for each
    /// element in the input.
    ///
    /// Runtime 12 ms Beats 75.00%
    /// Memory 2.49 MB Beats 85.14%
    pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
        let mut res = 0;
        let mut seen: HashMap<i32, i32> = HashMap::with_capacity(nums.len());
        seen.insert(0, 1);
        let mut prefix_sum = 0;
        let mut diff;
        for num in nums {
            prefix_sum += num;
            diff = prefix_sum - k;
            if let Some(entry) = seen.get(&diff) {
                res += entry;
            }
            seen.entry(prefix_sum)
                .and_modify(|counter| *counter += 1)
                .or_insert(1);
        }
        res
    }
}

// Tests.
fn main() {
    let tests = [(vec![1, 1, 1], 2, 2), (vec![1, 2, 3], 3, 2)];
    for t in tests {
        assert_eq!(Solution::subarray_sum(t.0, t.1), t.2);
    }
    println!("\x1b[92m» All tests passed!\x1b[0m")
}
